File: src/modules/payment/payment.service.ts
import { randomUUID } from 'node:crypto';
import { env } from '../../config/env.js';
import { AppError } from '../../shared/errors/AppError.js';
import { recordAudit } from '../../shared/utils/audit.js';
import { Order, type OrderDocument } from '../order/order.model.js';
import * as orderService from '../order/order.service.js';
import { User } from '../user/user.model.js';
import {
  Payment,
  type IPaymentRefund,
  type PaymentDocument,
  type PaymentStatus,
} from './payment.model.js';
import {
  getPaymentProvider,
  registerPaymentProvider,
  type WebhookEvent,
} from './payment.provider.js';
import { MockPaymentProvider } from './payment.providers/mock.provider.js';

// The mock provider is the default gateway. Importing this module registers it
// in the provider registry, so any code path that touches payments gets a
// working provider.
registerPaymentProvider(new MockPaymentProvider());

export interface PaymentPublic {
  id: string;
  orderId: string;
  userId: string;
  provider: string;
  providerReference: string;
  amountCents: number;
  currency: string;
  status: PaymentStatus;
  refundedCents: number;
  refunds: Array<{ amountCents: number; reason: string; refundReference: string; at: string }>;
  createdAt: string;
  updatedAt: string;
}

function toPaymentPublic(payment: PaymentDocument): PaymentPublic {
  return {
    id: payment._id.toString(),
    orderId: payment.orderId.toString(),
    userId: payment.userId.toString(),
    provider: payment.provider,
    providerReference: payment.providerReference,
    amountCents: payment.amountCents,
    currency: payment.currency,
    status: payment.status,
    refundedCents: payment.refundedCents,
    refunds: payment.refunds.map((r) => ({
      amountCents: r.amountCents,
      reason: r.reason,
      refundReference: r.refundReference,
      at: r.at.toISOString(),
    })),
    createdAt: payment.createdAt.toISOString(),
    updatedAt: payment.updatedAt.toISOString(),
  };
}

// ---------------------------------------------------------------------------
// Checkout
// ---------------------------------------------------------------------------

export interface CheckoutResult {
  payment: PaymentPublic;
  order: OrderPublic;
}

// Starts (or resumes) payment for a pending order. Safe to call repeatedly:
//  - an in-flight payment attempt (status 'created') is reused, not duplicated;
//  - a failed attempt re-reserves the released stock and starts a fresh one;
//  - the provider-level idempotency key + unique index stop double inserts.
export async function createCheckout(
  orderId: string,
  userId: string,
  opts?: { provider?: string },
): Promise<CheckoutResult> {
  const order = await Order.findById(orderId);
  if (!order || order.userId.toString() !== userId) throw AppError.notFound('Order not found');
  if (order.status !== 'pending') {
    throw AppError.conflict('Only pending orders can be checked out', {
      errorCode: 'ORDER_NOT_PAYABLE',
    });
  }
  if (order.paymentStatus === 'failed') {
    await orderService.prepareOrderForPayment(orderId);
  }

  let payment = await Payment.findOne({ orderId, status: { $in: ['created'] } }).sort({
    createdAt: -1,
  });
  if (!payment) {
    const user = await User.findById(userId).lean();
    const provider = getPaymentProvider(opts?.provider ?? env.PAYMENT_PROVIDER);
    const idempotencyKey = `ck_${orderId}_${randomUUID()}`;
    const result = await provider.createPayment({
      idempotencyKey,
      orderId,
      customer: {
        id: userId,
        email: user?.email ?? '',
        name: user?.name ?? '',
      },
      amountCents: order.totalCents,
      currency: order.currency,
    });

    try {
      payment = await Payment.create({
        idempotencyKey,
        orderId,
        userId,
        provider: provider.name,
        providerReference: result.providerReference,
        amountCents: order.totalCents,
        currency: order.currency,
        status: 'created',
        metadata: result.raw,
      });
    } catch (error) {
      // Unique index collision on idempotencyKey (concurrent double-submit of
      // the same checkout) — load the winner and proceed with it.
      if ((error as { code?: number }).code === 11000) {
        payment = await Payment.findOne({ idempotencyKey }).sort({ createdAt: -1 });
        if (!payment) throw AppError.internal('Checkout conflict could not be resolved');
      } else {
        throw error;
      }
    }

    if (result.status === 'succeeded') {
      await applyPaymentSucceeded(payment);
    }
  }

  return { payment: toPaymentPublic(payment), order: toOrderPublicSafe(order) };
}

// ---------------------------------------------------------------------------
// Webhook handling (raw body + HMAC signature)
// ---------------------------------------------------------------------------

export interface WebhookResult {
  acknowledged: boolean;
  handled: boolean;
  duplicate?: boolean;
}

// Called by the raw-body webhook route. `providerName` comes from the URL, so a
// single endpoint can serve multiple gateways. All side effects are applied
// through `handleEvent`, which is the single place order/payment state changes.
export async function processWebhook(
  providerName: string,
  rawBody: Buffer,
  signature: string,
): Promise<WebhookResult> {
  const provider = getPaymentProvider(providerName);
  const event = await provider.verifyWebhook(rawBody, signature);

  const payment = await Payment.findOne({
    provider: providerName,
    providerReference: event.providerReference,
  });
  if (!payment) {
    // The provider keeps retrying unacknowledged events, so acknowledge
    // orphans to stop the retry loop. Log them for manual reconciliation.
    recordAudit('payment.webhook_orphan', {
      providerReference: event.providerReference,
      type: event.type,
    });
    return { acknowledged: true, handled: false };
  }

  // Reject anything that does not match the payment we stored (tampering /
  // misrouted webhook). Refund events carry the refunded amount, not the
  // original charge, so they are exempt from the amount check.
  if (event.type !== 'refund.succeeded' && event.amountCents !== payment.amountCents) {
    throw AppError.badRequest('Webhook amount does not match the stored payment', {
      errorCode: 'WEBHOOK_AMOUNT_MISMATCH',
    });
  }

  if (payment.processedWebhookIds.includes(event.webhookId)) {
    return { acknowledged: true, handled: false, duplicate: true };
  }

  await handleEvent(payment, event);
  return { acknowledged: true, handled: true };
}

async function handleEvent(payment: PaymentDocument, event: WebhookEvent): Promise<void> {
  payment.processedWebhookIds.push(event.webhookId);
  switch (event.type) {
    case 'payment.succeeded':
      if (payment.status !== 'succeeded') {
        await applyPaymentSucceeded(payment);
      } else {
        await payment.save();
      }
      break;
    case 'payment.failed':
      if (payment.status === 'created') {
        await applyPaymentFailed(payment);
      } else {
        await payment.save();
      }
      break;
    case 'refund.succeeded':
      await applyRefundSucceeded(
        payment,
        event.amountCents,
        event.reason ?? '',
        event.refundReference ?? '',
      );
      break;
    default:
      throw AppError.badRequest('Unsupported webhook event type', {
        errorCode: 'UNSUPPORTED_WEBHOOK_EVENT',
      });
  }
}

// ---------------------------------------------------------------------------
// Side effects (the only place order + payment state mutate together)
// ---------------------------------------------------------------------------

async function applyPaymentSucceeded(payment: PaymentDocument): Promise<void> {
  await orderService.markOrderPaid(payment.orderId.toString(), payment._id.toString());
  payment.status = 'succeeded';
  await payment.save();
  recordAudit('payment.succeeded', {
    paymentId: payment._id.toString(),
    orderId: payment.orderId.toString(),
  });
}

async function applyPaymentFailed(payment: PaymentDocument): Promise<void> {
  await orderService.markOrderPaymentFailed(payment.orderId.toString());
  payment.status = 'failed';
  await payment.save();
  recordAudit('payment.failed', {
    paymentId: payment._id.toString(),
    orderId: payment.orderId.toString(),
  });
}

async function applyRefundSucceeded(
  payment: PaymentDocument,
  amountCents: number,
  reason: string,
  refundReference: string,
): Promise<void> {
  const refund: IPaymentRefund = { amountCents, reason, refundReference, at: new Date() };
  payment.refunds.push(refund);
  payment.refundedCents += amountCents;
  payment.status = payment.refundedCents >= payment.amountCents ? 'refunded' : 'partially_refunded';
  await payment.save();
  await orderService.finalizeRefund(
    payment.orderId.toString(),
    amountCents,
    reason,
    refundReference,
    'system',
  );
}

// ---------------------------------------------------------------------------
// Reads & refunds
// ---------------------------------------------------------------------------

export async function getPaymentByOrder(orderId: string, userId?: string): Promise<PaymentPublic> {
  const payment = await Payment.findOne({ orderId }).sort({ createdAt: -1 });
  if (!payment) throw AppError.notFound('Payment not found');
  if (userId && payment.userId.toString() !== userId) throw AppError.notFound('Payment not found');
  return toPaymentPublic(payment);
}

export async function refundOrder(
  orderId: string,
  input: { amountCents?: number; reason?: string },
  actorId: string,
): Promise<PaymentPublic> {
  const payment = await Payment.findOne({
    orderId,
    status: { $in: ['succeeded', 'partially_refunded'] },
  }).sort({ createdAt: -1 });
  if (!payment) throw AppError.notFound('No payable payment found for this order');

  const amountCents = input.amountCents ?? payment.amountCents - payment.refundedCents;
  if (amountCents <= 0) throw AppError.badRequest('Refund amount must be positive');

  const provider = getPaymentProvider(payment.provider);
  const result = await provider.refundPayment(payment.providerReference, {
    amountCents,
    reason: input.reason,
  });
  await applyRefundSucceeded(payment, amountCents, input.reason ?? '', result.refundReference);

  recordAudit('payment.refund_processed', { orderId, amountCents, actorId });
  return toPaymentPublic(payment);
}

// Order state used by the checkout response (imported lazily to avoid a
// circular import at module-evaluation time).
type OrderPublic = ReturnType<typeof orderService.toOrderPublic>;

function toOrderPublicSafe(order: OrderDocument): OrderPublic {
  return orderService.toOrderPublic(order);
}

File: java/src/main/java/com/ecommerce/backend/modules/payment/PaymentService.java
package com.ecommerce.backend.modules.payment;

import com.ecommerce.backend.common.error.ApiException;
import com.ecommerce.backend.config.AppConfig;
import com.ecommerce.backend.modules.audit.AuditService;
import com.ecommerce.backend.modules.order.Order;
import com.ecommerce.backend.modules.order.OrderRepository;
import com.ecommerce.backend.modules.order.OrderService;
import com.ecommerce.backend.modules.payment.Payment.PaymentRefund;
import com.ecommerce.backend.modules.payment.PaymentProvider.Registry;
import com.ecommerce.backend.modules.payment.PaymentProvider.WebhookEvent;
import com.ecommerce.backend.modules.user.User;
import com.ecommerce.backend.modules.user.UserRepository;
import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import org.springframework.dao.DuplicateKeyException;
import org.springframework.stereotype.Service;

/**
 * Checkout + webhook handling (mirrors {@code payment.service.ts}). All side
 * effects flow through the event handler below — the single place order and
 * payment state change together. Replays of gateway events are acknowledged but
 * ignored via {@code processedWebhookIds}.
 */
@Service
public class PaymentService {

    private final PaymentRepository paymentRepository;
    private final Registry providerRegistry;
    private final OrderRepository orderRepository;
    private final OrderService orderService;
    private final UserRepository userRepository;
    private final AuditService auditService;
    private final AppConfig config;

    public PaymentService(
            PaymentRepository paymentRepository,
            Registry providerRegistry,
            OrderRepository orderRepository,
            OrderService orderService,
            UserRepository userRepository,
            AuditService auditService,
            AppConfig config) {
        this.paymentRepository = paymentRepository;
        this.providerRegistry = providerRegistry;
        this.orderRepository = orderRepository;
        this.orderService = orderService;
        this.userRepository = userRepository;
        this.auditService = auditService;
        this.config = config;
    }

    // -------------------------------------------------------------- Public shapes

    public record PaymentRefundPublic(
            long amountCents, String reason, String refundReference, String at) {}

    public record PaymentPublic(
            String id,
            String orderId,
            String userId,
            String provider,
            String providerReference,
            long amountCents,
            String currency,
            String status,
            long refundedCents,
            List<PaymentRefundPublic> refunds,
            String createdAt,
            String updatedAt) {}

    public record CheckoutResult(PaymentPublic payment, OrderService.OrderPublic order) {}

    public record WebhookResult(boolean acknowledged, boolean handled, boolean duplicate) {}

    // --------------------------------------------------------------- Checkout

    public CheckoutResult createCheckout(String orderId, String userId) {
        Order order = orderRepository.findById(orderId)
                .orElseThrow(() -> ApiException.notFound("Order not found"));
        if (!userId.equals(order.userId)) {
            throw ApiException.notFound("Order not found");
        }
        if (!Order.STATUS_PENDING.equals(order.status)) {
            throw ApiException.conflict("Only pending orders can be checked out", "ORDER_NOT_PAYABLE");
        }
        if (Order.PAYMENT_FAILED.equals(order.paymentStatus)) {
            orderService.prepareOrderForPayment(orderId);
        }

        Payment payment = paymentRepository
                .findFirstByOrderIdAndStatusInOrderByCreatedAtDesc(
                        orderId, List.of(Payment.STATUS_CREATED))
                .orElse(null);
        if (payment == null) {
            User user = userRepository.findById(userId).orElse(null);
            PaymentProvider provider = providerRegistry.get(config.paymentProvider());
            String idempotencyKey = "ck_" + orderId + "_" + UUID.randomUUID();

            PaymentProvider.CreatePaymentResult result = provider.createPayment(
                    new PaymentProvider.CreatePaymentInput(
                            idempotencyKey,
                            orderId,
                            new PaymentProvider.Customer(userId, user == null ? "" : user.email, user == null ? "" : user.name),
                            order.totalCents,
                            order.currency));

            try {
                Payment created = new Payment();
                created.idempotencyKey = idempotencyKey;
                created.orderId = orderId;
                created.userId = userId;
                created.provider = provider.name();
                created.providerReference = result.providerReference();
                created.amountCents = order.totalCents;
                created.currency = order.currency;
                created.status = Payment.STATUS_CREATED;
                created.metadata = result.raw();
                created.createdAt = Instant.now();
                created.updatedAt = created.createdAt;
                payment = paymentRepository.insert(created);
            } catch (DuplicateKeyException e) {
                payment = paymentRepository
                        .findFirstByIdempotencyKeyOrderByCreatedAtDesc(idempotencyKey)
                        .orElseThrow(() -> ApiException.internal(
                                "Checkout conflict could not be resolved", "CHECKOUT_CONFLICT"));
            }

            if (Payment.STATUS_SUCCEEDED.equals(result.status())) {
                applyPaymentSucceeded(payment);
            }
        }

        // Re-read the order: auto-approval (or a retry after payment failure)
        // transitions it to confirmed, and the response must reflect that.
        Order refreshed = orderRepository.findById(orderId)
                .orElseThrow(() -> ApiException.notFound("Order not found"));
        return new CheckoutResult(toPaymentPublic(payment), orderService.toOrderPublic(refreshed));
    }

    // --------------------------------------------------------------- Webhook

    public WebhookResult processWebhook(String providerName, byte[] rawBody, String signature) {
        PaymentProvider provider;
        try {
            provider = providerRegistry.get(providerName);
        } catch (IllegalArgumentException e) {
            throw ApiException.badRequest(e.getMessage(), "UNKNOWN_PAYMENT_PROVIDER");
        }
        WebhookEvent event = provider.verifyWebhook(rawBody, signature);

        Payment payment = paymentRepository
                .findFirstByProviderAndProviderReference(providerName, event.providerReference())
                .orElse(null);
        if (payment == null) {
            auditService.log("payment.webhook_orphan", null, null, Map.of(
                    "providerReference", event.providerReference(),
                    "type", event.type()));
            return new WebhookResult(true, false, false);
        }

        if (!"refund.succeeded".equals(event.type()) && event.amountCents() != payment.amountCents) {
            throw ApiException.badRequest(
                    "Webhook amount does not match the stored payment", "WEBHOOK_AMOUNT_MISMATCH");
        }

        if (payment.processedWebhookIds.contains(event.webhookId())) {
            return new WebhookResult(true, false, true);
        }

        handleEvent(payment, event);
        return new WebhookResult(true, true, false);
    }

    private void handleEvent(Payment payment, WebhookEvent event) {
        payment.processedWebhookIds.add(event.webhookId());
        switch (event.type()) {
            case "payment.succeeded":
                if (!Payment.STATUS_SUCCEEDED.equals(payment.status)) {
                    applyPaymentSucceeded(payment);
                } else {
                    paymentRepository.save(payment);
                }
                break;
            case "payment.failed":
                if (Payment.STATUS_CREATED.equals(payment.status)) {
                    applyPaymentFailed(payment);
                } else {
                    paymentRepository.save(payment);
                }
                break;
            case "refund.succeeded":
                applyRefundSucceeded(
                        payment,
                        event.amountCents(),
                        event.reason() == null ? "" : event.reason(),
                        event.refundReference() == null ? "" : event.refundReference());
                break;
            default:
                throw ApiException.badRequest(
                        "Unsupported webhook event type", "UNSUPPORTED_WEBHOOK_EVENT");
        }
    }

    // ------------------------------------------------------------ Side effects

    private void applyPaymentSucceeded(Payment payment) {
        orderService.markOrderPaid(payment.orderId, payment.id);
        payment.status = Payment.STATUS_SUCCEEDED;
        payment.updatedAt = Instant.now();
        paymentRepository.save(payment);
        auditService.log("payment.succeeded", null, null, Map.of(
                "paymentId", payment.id, "orderId", payment.orderId));
    }

    private void applyPaymentFailed(Payment payment) {
        orderService.markOrderPaymentFailed(payment.orderId);
        payment.status = Payment.STATUS_FAILED;
        payment.updatedAt = Instant.now();
        paymentRepository.save(payment);
        auditService.log("payment.failed", null, null, Map.of(
                "paymentId", payment.id, "orderId", payment.orderId));
    }

    private void applyRefundSucceeded(
            Payment payment, long amountCents, String reason, String refundReference) {
        PaymentRefund refund = new PaymentRefund();
        refund.amountCents = amountCents;
        refund.reason = reason;
        refund.refundReference = refundReference;
        refund.at = Instant.now();
        payment.refunds.add(refund);
        payment.refundedCents += amountCents;
        payment.status = payment.refundedCents >= payment.amountCents
                ? Payment.STATUS_REFUNDED
                : Payment.STATUS_PARTIALLY_REFUNDED;
        payment.updatedAt = Instant.now();
        paymentRepository.save(payment);
        orderService.finalizeRefund(
                payment.orderId, amountCents, reason, refundReference, "system");
    }

    // --------------------------------------------------------------- Reads & refunds

    public PaymentPublic getPaymentByOrder(String orderId, String userId) {
        Payment payment = paymentRepository
                .findFirstByOrderIdOrderByCreatedAtDesc(orderId)
                .orElseThrow(() -> ApiException.notFound("Payment not found"));
        if (userId != null && !userId.equals(payment.userId)) {
            throw ApiException.notFound("Payment not found");
        }
        return toPaymentPublic(payment);
    }

    public PaymentPublic refundOrder(
            String orderId, Integer amountCents, String reason, String actorId) {
        Payment payment = paymentRepository
                .findFirstByOrderIdAndStatusInOrderByCreatedAtDesc(
                        orderId,
                        List.of(Payment.STATUS_SUCCEEDED, Payment.STATUS_PARTIALLY_REFUNDED))
                .orElseThrow(() -> ApiException.notFound("No payable payment found for this order"));

        long refundAmount = amountCents == null
                ? payment.amountCents - payment.refundedCents
                : amountCents;
        if (refundAmount <= 0) {
            throw ApiException.badRequest("Refund amount must be positive", "BAD_REQUEST");
        }

        PaymentProvider provider = providerRegistry.get(payment.provider);
        PaymentProvider.RefundResult result = provider.refundPayment(
                payment.providerReference, refundAmount, reason);
        applyRefundSucceeded(payment, refundAmount, reason == null ? "" : reason, result.refundReference());

        auditService.log("payment.refund_processed", actorId, null, Map.of(
                "orderId", orderId, "amountCents", refundAmount));
        return toPaymentPublic(payment);
    }

    // -------------------------------------------------------------- Internals

    private PaymentPublic toPaymentPublic(Payment payment) {
        List<PaymentRefundPublic> refunds = payment.refunds.stream()
                .map(r -> new PaymentRefundPublic(
                        r.amountCents, r.reason, r.refundReference, iso(r.at)))
                .toList();
        return new PaymentPublic(
                payment.id,
                payment.orderId,
                payment.userId,
                payment.provider,
                payment.providerReference,
                payment.amountCents,
                payment.currency,
                payment.status,
                payment.refundedCents,
                refunds,
                iso(payment.createdAt),
                iso(payment.updatedAt));
    }

    private static String iso(Instant value) {
        return value == null ? null : value.toString();
    }
}
